File: backend/app/api/alerts.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.alert import (
    AlertListResponse,
    AlertResponse,
    AlertAcknowledgeResponse,
    AlertActionResponse,
)
from app.services.alert_service import (
    list_alerts,
    acknowledge_alert,
    resolve_alert,
    escalate_alert,
)
from app.midlware.auth_middleware import get_current_nurse
from app.models.nurse import Nurse
from app.models.alert import Alert

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


@router.get("", response_model=AlertListResponse)
def get_alerts(
    status: str | None = Query(None),
    priority: str | None = Query(None),
    limit: int = Query(20),
    db: Session = Depends(get_db),
    nurse: Nurse = Depends(get_current_nurse),
):
    return list_alerts(db, status=status, priority=priority, limit=limit)


@router.get("/{alert_id}", response_model=AlertResponse)
def get_alert(alert_id: int, db: Session = Depends(get_db), nurse: Nurse = Depends(get_current_nurse)):
    alert = db.query(Alert).filter(Alert.id == alert_id).first()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    return alert


@router.post("/{alert_id}/acknowledge", response_model=AlertAcknowledgeResponse)
def acknowledge(alert_id: int, db: Session = Depends(get_db), nurse: Nurse = Depends(get_current_nurse)):
    result = acknowledge_alert(db, alert_id, nurse.id)
    if not result:
        raise HTTPException(status_code=400, detail="Alert not found or already acknowledged")
    return result


@router.post("/{alert_id}/resolve", response_model=AlertActionResponse)
def resolve(alert_id: int, db: Session = Depends(get_db), nurse: Nurse = Depends(get_current_nurse)):
    result = resolve_alert(db, alert_id, nurse.id)
    if not result:
        raise HTTPException(status_code=400, detail="Alert not found or not in progress")
    return result


@router.post("/{alert_id}/escalate", response_model=AlertActionResponse)
def escalate(alert_id: int, db: Session = Depends(get_db), nurse: Nurse = Depends(get_current_nurse)):
    result = escalate_alert(db, alert_id)
    if not result:
        raise HTTPException(status_code=400, detail="Alert not found")
    return result

File: backend/src/routes/alerts.ts
import { Router } from 'express'
import { authRequired, type AuthRequest } from '../middleware/auth.js'
import {
  getAlerts,
  getAlertById,
  acknowledgeAlert,
  resolveAlert,
  escalateAlert,
} from '../services/alert.js'

const router = Router()

router.get('/', authRequired, async (req: AuthRequest, res) => {
  try {
    const { status, priority, limit } = req.query
    const alerts = await getAlerts({
      status: status as string | undefined,
      priority: priority as string | undefined,
      limit: limit ? parseInt(limit as string, 10) : undefined,
    })
    const high = alerts.filter(a => a.risk_priority === 'high').length
    const medium = alerts.filter(a => a.risk_priority === 'medium').length
    const low = alerts.filter(a => a.risk_priority === 'low').length
    res.json({ alerts, total: alerts.length, high, medium, low })
  } catch (err) {
    console.error('[ALERTS] List error:', err)
    res.status(500).json({ error: 'Internal server error' })
  }
})

router.get('/:id', authRequired, async (req: AuthRequest, res) => {
  try {
    const id = parseInt(req.params.id as string, 10)
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid alert ID' })
      return
    }
    const alert = await getAlertById(id)
    if (!alert) {
      res.status(404).json({ error: 'Alert not found' })
      return
    }
    res.json(alert)
  } catch (err) {
    console.error('[ALERTS] Detail error:', err)
    res.status(500).json({ error: 'Internal server error' })
  }
})

router.post('/:id/acknowledge', authRequired, async (req: AuthRequest, res) => {
  try {
    const id = parseInt(req.params.id as string, 10)
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid alert ID' })
      return
    }
    const assignment = await acknowledgeAlert(id, req.user!.userId)
    res.json({
      id,
      status: 'in_progress',
      assigned_to: { id: req.user!.userId },
      acknowledged_at: assignment.accepted_at,
    })
  } catch (err: any) {
    if (err.message === 'Alert already acknowledged by another nurse') {
      res.status(409).json({ error: err.message })
      return
    }
    console.error('[ALERTS] Acknowledge error:', err)
    res.status(500).json({ error: 'Internal server error' })
  }
})

router.post('/:id/resolve', authRequired, async (req: AuthRequest, res) => {
  try {
    const id = parseInt(req.params.id as string, 10)
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid alert ID' })
      return
    }
    await resolveAlert(id, req.user!.userId)
    res.json({ id, status: 'resolved' })
  } catch (err) {
    console.error('[ALERTS] Resolve error:', err)
    res.status(500).json({ error: 'Internal server error' })
  }
})

router.post('/:id/escalate', authRequired, async (req: AuthRequest, res) => {
  try {
    const id = parseInt(req.params.id as string, 10)
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid alert ID' })
      return
    }
    await escalateAlert(id, req.user!.userId)
    res.json({ id, status: 'escalated' })
  } catch (err) {
    console.error('[ALERTS] Escalate error:', err)
    res.status(500).json({ error: 'Internal server error' })
  }
})

export default router
